Qt: Setting for deciding address type (legacy, p2sh or bech32)

hsjoberg opened the pull request:
Related PR #11403, needs to be merged first. I'll rebase this PR to be up to date with #11403, and rebase it to master once it has been merged.

Exposes the `addresstype` parameter through the GUI.
Currently directly changes the `g_address_type` global variable instead of requiring an application restart, I'm not sure if this is good practice, but could easily be changed if needed.

Radio buttons seem to be acting a bit different compared to other GUI elements, so that's why the code is somewhat different than for other settings.

hsjoberg committed:
Qt: Adding address type GUI options

The Options-dialog now contains radio buttons for deciding which address
type to use (legacy, p2sh or BIP173 (bech32)).

This setting is connected to the `addresstype` parameter.

Sjors commented:
If you edit the pull request description, you can change the base branch from `bitcoin:master` to `sipa:201709_segwitwallet2`. That should make review here much easier. Although it might cause Github to move the whole pull request to @sipa's account, which would hurt visibility.

hsjoberg commented:
@Sjors Yes, I tried to do that but it doesn't seem to be possible.

Sjors commented:
No problem, it's easy to work around:
<img width="712" alt="schermafbeelding 2017-12-18 om 18 23 23" src="https://user-images.githubusercontent.com/10217/34118906-94c3fa46-e420-11e7-8769-ef3d993ab5ec.png">

Sjors reviewed (commented):
Concept ACK. It's great to offer users the ability to use bech32 addresses.

Receive tab works as expected.

Send tab has a bug: without a restart, the change address will not use newly selected address type (maybe explicitly set `-changetype`?).

The nested boxes and alignment is not very pretty:

<img width="691" alt="nesting" src="https://user-images.githubusercontent.com/10217/34120688-cd831e06-e426-11e7-94ef-3db971522d67.png">

I suggest moving this the same level as Expert. Alternatively, use a dropdown.

Arguably this isn't really an expert feature anyway, especially if you remove the Legacy option. Users can't really shoot themselves in the foot: either they get paid, or the creditor tells them they don't understand the address, in which case the user creates a P2SH address.

I'd like to have a way to toggle this in the address generation screen too, but that can be done independently of this PR.

promag commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -163,6 +163,57 @@
            </widget>
           </item>
+          <item>
+           <widget class="QGroupBox" name="groupBoxAddressType">
```
This is inside Export group box, should be on the outside instead?

hsjoberg replied:
Fixing.

hsjoberg commented:
> I suggest moving this the same level as Expert. Alternatively, use a dropdown.

@Sjors @promag The Address type was actually intentionally in the Expert group, but I guess you are right, it makes more sense to have it separately. I will move it.
We need good tooltips and description though.

hsjoberg commented on src/qt/optionsdialog.cpp:
```diff
@@ -180,6 +181,10 @@ void OptionsDialog::setMapper()
     mapper->addMapping(ui->spendZeroConfChange, OptionsModel::SpendZeroConfChange);
     mapper->addMapping(ui->coinControlFeatures, OptionsModel::CoinControlFeatures);
 
+    mapper->addMapping(ui->addressTypeLegacy, OptionsModel::AddressTypeLegacy);
```
Hmm, I think it makes more sense to have the GUI options mapped 1:1 to the `addresstype` parameter. Perhaps this could be something to look into later?

Sjors replied:
Nvm, it's fine for a setting, especially combined with the per address checkbox in #11991.

hsjoberg commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -163,6 +163,57 @@
+                </sizepolicy>
+               </property>
+               <property name="text">
+                <string>BIP173 (bech32)</string>
```
Sure, I'll look into that.
Any suggestions of exactly what the tooltip(s) should say?

hsjoberg commented on src/qt/optionsdialog.cpp:
```diff
@@ -329,6 +334,23 @@ void OptionsDialog::updateDefaultProxyNets()
+        break;
+    case OUTPUT_TYPE_NONE:
+    default:
+        break;
```
Hmm I'm not sure if we really need to kill the application, if we cannot resolve an address type, it doesn't necessarily need to be fatal.

I'm not sure what the best practice is here.

jonasschnelli replied:
I think an assert would be appropriate here (stop the app if address type is unknown to the GUI).

promag replied:
@hsjoberg ping.

hsjoberg commented on src/qt/optionsmodel.cpp:
```diff
@@ -110,6 +110,11 @@ void OptionsModel::Init(bool resetSettings)
     if (!gArgs.SoftSetBoolArg("-spendzeroconfchange", settings.value("bSpendZeroConfChange").toBool()))
         addOverriddenOption("-spendzeroconfchange");
+
+    if(!settings.contains("strAddressType"))
```
Thanks, fixing missing space. :)

hsjoberg commented:
> Send tab has a bug: without a restart, the change address will not use newly selected address type (maybe explicitly set -changetype?).

@Sjors Very true, I'll fix this.

hsjoberg commented:
Fixed some of the nits.

luke-jr commented:
GUI settings probably shouldn't influence RPC like this, and IMO it should really be exposed per-receive.

Sjors commented:
@luke-jr wrote:
>  IMO it should really be exposed per-receive

Do you mean that it should _only_ be exposed per receive (i.e. don't add this setting)?

That would not be ideal, because the wallet wouldn't use bech32 for change in that case.

It reveals another problem with this setting: if the user changes the setting and then runs `bitcoind` instead of `bitcoin-qt`, they might expect the RPC to use bech32 as well, just like it would in the debug console inside of QT.

One counter argument to that is that we expect `bitcoind` users to know what they're doing and to aware of the fact that GUI settings, like "spend unconfirmed change",  don't apply to `bitcoind`. This would be worth pointing out in the release notes and such.

There's something to be said for being able to persist wallet settings, so that they apply to both `bitcoin-qt` and `bitcoind` if no flag overrides it.

Writing the setting into bitcoin.conf is one way, but at least one potential problem with that, is that we can't change the default later, because all users have it set in stone in their bitcoin.conf file. That's probably an issue with the PR as well; once there is a setting, it's weird to just change it.

So I might agree with @luke-jr here in that this should be a per-receive thing, and not a setting. But then I would  suggest that we make the change address behavior more intelligent (not in this PR): its address type should match the destination. Also making it a per receive checkbox still makes it difficult to change, as the user might grow accustomed to the check-box being checked or unchecked.

So I'm not really sure. We should think about how to encourage users to try out bech32  without making it more difficult to make this the default in the future.

hsjoberg commented:
> Send tab has a bug: without a restart, the change address will not use newly selected address type (maybe explicitly set -changetype?).

@Sjors Rebased and fixed this.
Right now, changetype isn't changed if you explicitly set a type it in the configuration file, otherwise it defaults to whatever addresstype option is chosen.

hsjoberg commented:
> GUI settings probably shouldn't influence RPC like this

@luke-jr Isn't this what GUI settings basically are used for? It seems like most of the settings in the GUI exposes different configuration parameters.

> IMO it should really be exposed per-receive.

Yes, I guess that makes sense too.

Sjors commented:
For a per receive address toggle, see #11991. These PR's are compatible (aside from the usual potential for merge conflicts), so we can use both, either or none.

jb55 commented:
Hmm after testing this patch over #11991, I think I prefer this one. The toggle seems a bit weird. I don't see why I would ever want to selectively toggle a bech32 address, vs just eventually having it my default when I feel comfortable doing that.

sipa commented:
@jb55 The only reason for bech32 or not is whether the sender supports bech32 - which I think is something you may want to decide per address.

hsjoberg commented:
Perhaps it makes sense to have both options available. Per receive level and as a global setting.
If not, I think @Sjors PR is works very well.

molxyz commented:
ACK!  I just compiled this PR for Windows. It is beautiful. Really love it.  Will test. 

![walletgui-by-coco](https://user-images.githubusercontent.com/21127776/34495853-0a1e972e-efc5-11e7-8cf4-10a1cec80a21.png)

jonasschnelli commented on src/qt/optionsmodel.cpp:
```diff
@@ -381,6 +386,27 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
+        case AddressTypeLegacy:
+            if (value.toBool()) {
+                settings.setValue("strAddressType", QString::fromStdString(FormatOutputType(OUTPUT_TYPE_LEGACY)));
+                g_address_type = OUTPUT_TYPE_LEGACY;
```
Not sure, but I guess changing `g_address_type` and `g_change_type ` needs locking (concurrency), otherwise a race would be possible (unlikely, though possible).

promag replied:
Guard with cs_main? Most accesses already held the lock. Fix the remaining.

jonasschnelli commented:
Nice!
Concept ACK

jonasschnelli commented:
Needs rebase

hsjoberg commented:
@jonasschnelli I'll rebase to master as soon as I have time.

hsjoberg commented:
Rebased to master.
I intend to look at and fix the things that have been pointed out soon.

jonasschnelli commented on src/qt/optionsdialog.cpp:
```diff
@@ -329,6 +334,23 @@ void OptionsDialog::updateDefaultProxyNets()
+    case OUTPUT_TYPE_LEGACY:
+        ui->addressTypeLegacy->setChecked(true);
+        break;
+    case OUTPUT_TYPE_P2SH:
```
Rebase issue, .. should this be `OUTPUT_TYPE_P2SH_SEGWIT`?

Sjors replied:
Yes, it won't compile otherwise.

hsjoberg replied:
@jonasschnelli Fixed, thanks!

jonasschnelli commented on src/qt/optionsmodel.h:
```diff
@@ -48,6 +49,9 @@ class OptionsModel : public QAbstractListModel
         DatabaseCache,          // int
         SpendZeroConfChange,    // bool
+        AddressTypeLegacy,      // bool
+        AddressTypeP2SH,        // bool
```
Maybe for constancy reason, change that to `AddressTypeP2SHSegWit`?

promag replied:
Agree.

hsjoberg replied:
Fixed.

Sjors commented on src/qt/optionsmodel.cpp:
```diff
@@ -110,6 +110,11 @@ void OptionsModel::Init(bool resetSettings)
+    if (!settings.contains("strAddressType"))
+      settings.setValue("strAddressType", QString::fromStdString(FormatOutputType(OUTPUT_TYPE_DEFAULT)));
+    if (!gArgs.SoftSetArg("-addresstype", settings.value("strAddressType").toString().toStdString()))
+        addOverriddenOption("-addresstype");
```
Actually this is not true. It will happily ignore `-addresstype` if you change the setting.

Instead, you should add `addOverriddenOption("-changetype");`, since this setting _is_ honored.

Sjors commented on src/qt/optionsmodel.h:
```diff
@@ -48,6 +49,9 @@ class OptionsModel : public QAbstractListModel
         SpendZeroConfChange,    // bool
+        AddressTypeLegacy,      // bool
+        AddressTypeP2SH,        // bool
+        AddressTypeBIP173,      // bool
```
`AddressTypeBech32` would be more consistent with `enum OutputType`.

promag replied:
Agree.

hsjoberg replied:
Fixed.

promag commented on src/qt/optionsmodel.cpp:
```diff
@@ -381,6 +386,27 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
             }
             break;
+        case AddressTypeLegacy:
+            if (value.toBool()) {
```
Looks like this code could be factored out to a function? (repeated 3 times)

instagibbs commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -166,6 +166,57 @@
+             </sizepolicy>
+            </property>
+            <property name="text">
+             <string>P2SH</string>
```
prefered(merged) nomenclature is "P2SH-Segwit"

hsjoberg replied:
@instagibbs Fixed.

hsjoberg commented:
Addressed some of the issues that have been pointed out as well as the compiling error, sorry for that.

There are still some outstanding things that need to be addressed, I'll look into them.

jonasschnelli commented on src/qt/optionsmodel.cpp:
```diff
@@ -110,6 +110,11 @@ void OptionsModel::Init(bool resetSettings)
         addOverriddenOption("-spendzeroconfchange");
+
+    if (!settings.contains("strAddressType"))
+        settings.setValue("strAddressType", QString::fromStdString(FormatOutputType(OUTPUT_TYPE_DEFAULT)));
```
We do that also at different places... but this makes changing default in the future really hard. If we the default for `-addresstype`, QT remains at the old state even if the user hasn't changed the value there...

jonasschnelli replied:
We can fix this in a follow up PR (that also fixes the other issues with defaults).

jonasschnelli commented:
Tested a bit and seems to work.
What worries me is that the code has not the right concurrency protections to have those globals change during run-time. There would be some potential races (rare and often only possible when mixing QT with RPC).

We can leave this for an follow up PR but I'd prefer the have the right instruments in place before adding runtime "sets" on those globals.

gmaxwell commented:
I don't see any use case for the three way selection-- it just confronts users with more technical details (complete with gobbltygopy technical terms like P2SH and BIP173 ) which many won't know how to select among.

The purpose of having the legacy setting is maintaining wallet backwards compatibility.  .. but that's blown out the moment you select either P2SH or BC1 addresses.   What is the envisioned use-case for someone to set legacy but then sometimes pick P2SH?   Without understanding that I am a Concept NAK.

Being able to switch between BC1 and 3xxx addresses at receive time would _very_ useful, but that should be a per-receive selection set when the address is requested.

jonasschnelli commented:
Removing 0.16 milestone since there are some concerns with this PR. Also, seems not to be a blocker for 0.16 (especially since #11991).

molxyz commented:
I'm not sure why this PR isn't merged? It makes much more sense to use this so I am using it as my wallet. Please consider to merge this PR.  Thank you!

MarcoFalke commented:
Needs rebase if still relevant. Otherwise I suggest to close.

hsjoberg commented:
@molxyz @MarcoFalke Sorry for the delay.

Yes, If there is interest I could continue work on this PR.

There are some minor outstanding issues (albeit minor) regarding thread safety and other things that I'll look into before merge.

There doesn't seem to be a clear consensus regarding if we should merge this.
Personally I think it makes sense considering we already provide these options through the configuration file and as a command-line argument. I do understand the argument that it could be perceived as too technical though (not sure on how to improve on this?)

MarcoFalke commented:
@hsjoberg have you seen #13251? I haven't looked closely at you pull request, but they seem strongly related. And yours might not be relevant after that one has been merged.

MarcoFalke commented:
If it is still relevant, please rebase so that it compiles at a minimum. (green travis)

MarcoFalke commented on src/qt/optionsmodel.h:
```diff
@@ -6,6 +6,7 @@
 #define BITCOIN_QT_OPTIONSMODEL_H
 
 #include <amount.h>
+#include <wallet/wallet.h>
```
Why this include?

MarcoFalke commented:
Hmm, I don't think we should expose a "legacy" checkbox through the gui. I am missing the use case for that, given that "bech32" can already be unchecked.

practicalswift commented:
This PR doesn't compile when rebased on `master` :-)

MarcoFalke commented:
Closing for now
